### PR TITLE
FEATURE: EGL-364 introduce blacklist for attribute hierarchies

### PR DESCRIPTION
**Pull request by phong-nguyen-duy**
---

Supported PR commands:

| Command                                                 | Description                                                |
| ------------------------------------------------------- | ---------------------------------------------------------- |
| `ok to test`                                            | Re-run standard checks                                     |
| `extended check sonar`                                  | SonarQube tests                                            |
| `extended test - backstop`                              | BackstopJS tests                                           |
| **E2E Cypress tests commands - TIGER**                  |                                                            |
| `extended test - tiger-cypress - isolated <testName>`   | Run isolated tests running against recorded Tiger backend. |
| `extended test - tiger-cypress - record <testName>`     | Create a new recording for isolated Tiger tests.           |
| `extended test - tiger-cypress - integrated <testName>` | Run integrated tests against live backend                  |
| **E2E Cypress tests commands - BEAR**                   |                                                            |
| `extended test - cypress - isolated <testName>`         | Run isolated tests running against recorded Bear backend.  |
| `extended test - cypress - record <testName>`           | Create a new recording for isolated Bear tests.            |
| `extended test - cypress - integrated <testName>`       | Run integrated tests against live backend                  |
| **Compatibility matrix test commands - TIGER Backend**  |                                                            |
| `extended test - matrix-test <AIO_version>`             | Run integrated tests against AIO versions.                 |

`<testName>` in cypress commands is used to filter specfiles. Example, to run record with BEAR backend

-   Against `dashboard.spec.ts` and `drilling.spec.ts`, execute command `extended test - cypress - record dashboard,drilling`
-   Against all specfiles, execute command `extended test - cypress - record` or `extended test - cypress - record *`

`<AIO_version>` in commands is used to start test with multiple AIO instances - each instance in triggered by one jenkins build

-   To run with `master` and `stable`, execute command `extended test - matrix-test master,stable` or `extended test - matrix-test latest`
-   To run with specific version,ex: `2.3.0` and `2.3.1`, execute command `extended test - matrix-test 2.3.0,2.3.1`
-   In case `<AIO_version>` is empty, read versions from file `compTigerVersions.txt` of this repo

---

# PR Checklist

-   [ ] commit messages adhere to the [commit message guidelines](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#what-should-the-commits-look-like)
-   [ ] review was done by a Code owner [if necessary](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#how-do-i-tell-if-my-pull-request-needs-approval-by-a-code-owner) (if you think it is not necessary, explain the reasoning in the description or in a comment)
-   [ ] `check` passes
-   [ ] `extended test - backstop` passes
-   [ ] `extended test - tiger-cypress - record` to record new mapping files (Tiger BE)
-   [ ] `extended test - cypress - record` to record new mapping files (Bear BE)
-   [ ] `extended test - tiger-cypress - isolated` passes
-   [ ] `extended test - cypress - isolated` passes
-   [ ] `extended test - tiger-cypress - integrated` passes
-   [ ] `rush change` [was run if applicable](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#how-do-i-describe-my-changes-for-the-changelog)

### CODE REVIEW

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/16846/ : SUCCESS in 3s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/18272/ : SUCCESS in 13m 04s

File: libs/sdk-ui-dashboard/src/model/store/catalog/catalogSelectors.ts [hunk 1]
```diff
@@ -281,3 +293,58 @@ export const selectCatalogDateAttributeToDataset: DashboardSelector<
...
+): Record<string, ObjRef[]> {
+    const allCatalogAttributes = [...attributes, ...dateAttributes];
+    const attributeDescendants: Record<string, ObjRef[]> = {};
+
+    allCatalogAttributes.forEach((attribute) => {
+        const attributeRef = attribute.attribute.ref;
+        attributeHierarchies.forEach((hierarchy) => {
+            const hierarchyAttributes = hierarchy.attributeHierarchy.attributes.filter((attrRef) => {
+                const ignoredIndex = ignoredDrillDownHierarchies.findIndex((reference) => {
+                    if (isAttributeHierarchyReference(reference)) {
```
Comment (xMort) [inline libs/sdk-ui-dashboard/src/model/store/catalog/catalogSelectors.ts:311]: Please extract this utility block somewhere. It is repeated multiple times in the pull code. (311-323 lines)

File: libs/sdk-ui-dashboard/src/model/store/widgetDrills/widgetDrillSelectors.ts [hunk 2]
```diff
@@ -199,7 +206,20 @@ function getGlobalDrillDownAttributeHierarchyDefinitions(
                 const isAttributeInHierarchy = attributeIdentifiers.includes(
                     objRefToString(attributeHeader.formOf.ref),
                 );
-                if (isAttributeInHierarchy) {
+                const inBlacklistIndex = ignoredDrillDownHierarchies.findIndex((reference) => {
+                    if (isAttributeHierarchyReference(reference)) {
```
Comment (xMort) [inline libs/sdk-ui-dashboard/src/model/store/widgetDrills/widgetDrillSelectors.ts:210]: Duplicated.

File: libs/sdk-ui-dashboard/src/presentation/widget/insight/configuration/DrillTargetType/DrillTargetTypeListItem.tsx [hunk 3]
```diff
@@ -25,12 +29,23 @@ const DrillTargetTypeListItem: React.FC<IDrillTargetTypeListItemProps> = ({
...
+                className={className}
+                title={item.title}
+                isSelected={isSelected}
+                onClick={handleClick}
+            />
+            {item.disabled && item.disableTooltipMessage ? (
+                <Bubble
+                    className="bubble-primary"
+                    alignPoints={[{ align: "cr cl" }]}
+                    arrowOffsets={{ "cr cl": [15, 0] }}
```
Comment (xMort) [inline libs/sdk-ui-dashboard/src/presentation/widget/insight/configuration/DrillTargetType/DrillTargetTypeListItem.tsx:43]: Please, extract these to a constants outside of this component.

File: libs/sdk-ui-dashboard/src/presentation/widget/insight/configuration/DrillTargets/AttributeHierarchyListItem.tsx [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+export const AttributeHierarchyListItem: React.FC<IAttributeHierarchyListItemProps> = (props) => {
+    const { onClick, item } = props;
+    const catalogAttributes = useDashboardSelector(selectCatalogAttributes);
+    const catalogDateAttributes = useDashboardSelector(selectCatalogDateAttributes);
+
+    const hierarchyListItemClassname = cx(
+        "attribute-hierarchy-list-item s-attribute-hierarchy-list-item",
+        `s-${stringUtils.simplifyText(item.attributeHierarchy.title)}`,
+    );
+    const attributeDetailItems = item.attributeHierarchy.attributes
```
Comment (xMort) [inline libs/sdk-ui-dashboard/src/presentation/widget/insight/configuration/DrillTargets/AttributeHierarchyListItem.tsx:38]: Please extract this mapper to a function outside of this component to make it more readable.

File: libs/sdk-ui-dashboard/src/model/commandHandlers/widgets/addDrillDownForInsightWidgetHandler.ts [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        payload: { attributeIdentifier, attributeHierarchy },
+        correlationId,
+    } = cmd;
+    const widgets: ReturnType<typeof selectWidgetsMap> = yield select(selectWidgetsMap);
+    const insightWidget = validateExistingInsightWidget(widgets, cmd, ctx);
+    const { ref: widgetRef, ignoredDrillDownHierarchies: currentBlacklistHierarchies } = insightWidget;
+
+    const newBlacklistHierarchies = currentBlacklistHierarchies
+        ? currentBlacklistHierarchies.filter((ref) => {
+              if (isAttributeHierarchyReference(ref)) {
```
Comment (xMort) [inline libs/sdk-ui-dashboard/src/model/commandHandlers/widgets/addDrillDownForInsightWidgetHandler.ts:34]: Duplicated.

File: libs/sdk-ui-dashboard/src/presentation/widget/insight/configuration/DrillTargets/DrillTargetAttributeHierarchyItem.tsx [hunk 6]
```diff
@@ -1,29 +1,126 @@
...
+                    areObjRefsEqual(ref.dateHierarchyTemplate, hierarchy.attributeHierarchy.ref) &&
+                    objRefToString(ref.dateDatasetAttribute) ===
+                        attributeDescriptor?.attributeHeader.identifier
+                );
+            }
+        });
+    });
+
+    const emptyHierarchyMessage = intl.formatMessage(messages.emtpyHierarchyInfo);
+    const shouldShowEmptyMessage = isEmpty(existBlacklistHierarchies) && !config.complete;
```
Comment (khanhhuynh-gooddata) [inline libs/sdk-ui-dashboard/src/presentation/widget/insight/configuration/DrillTargets/DrillTargetAttributeHierarchyItem.tsx:82]: In cases where the origin attribute exists in more than one attribute hierarchy, displaying an empty message does not make sense. I would suggest only showing the empty message if the origin attribute is not included in any attribute hierarchy. In such cases, the hierarchy dropdown should be shown, but the item not in the blacklist will be disabled.

Could you help check, thanks!

File: libs/sdk-ui-dashboard/src/presentation/widget/insight/configuration/DrillTargets/AttributeHierarchyListItem.tsx [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+        })
+        .filter((item): item is IAttributeHierarchyDetailItem => !!item);
+    return (
+        <div className={hierarchyListItemClassname} onClick={onClick}>
+            <div className="attribute-hierarchy-list-item-content s-attribute-hierarchy-list-item-content">
+                <ShortenedText>{item.attributeHierarchy.title}</ShortenedText>
+            </div>
+
+            <div className="attribute-hierarchy-list-item-description s-attribute-hierarchy-list-item-description">
+                <AttributeHierarchyDetailBubble>
```
Comment (khanhhuynh-gooddata) [inline libs/sdk-ui-dashboard/src/presentation/widget/insight/configuration/DrillTargets/AttributeHierarchyListItem.tsx:66]: Is there any reason why we separate the detail component into two parts?

File: libs/sdk-ui-ext/styles/internal/scss/attribute_hierarchy_detail.scss [hunk 8]
```diff
@@ -0,0 +1,55 @@
+// (C) 2023 GoodData Corporation
+
+@use "@gooddata/sdk-ui-kit/styles/scss/variables" as kit-variables;
+
+.gd-attribute-hierarchy-detail-panel {
+    width: 200px;
+    height: 170px;
```
[the remark this answers is not in the document]
Comment (phong-nguyen-duy) [reply libs/sdk-ui-ext/styles/internal/scss/attribute_hierarchy_detail.scss]: Removed

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/16864/ : SUCCESS in 3s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/18294/ : FAILURE in 3m 04s

File: libs/sdk-ui-dashboard/src/model/store/catalog/catalogSelectors.ts (hunk 1 above)
Comment (phong-nguyen-duy) [reply libs/sdk-ui-dashboard/src/model/store/catalog/catalogSelectors.ts]: Extracted to [attributeHierarchyUtils.ts](https://github.com/gooddata/gooddata-ui-sdk/pull/4339/files#diff-9f3c4d8d04ae077a880336f4b79fb262cd1867580d690411c91f460be18ee254) and replaced all duplicated code

File: libs/sdk-ui-dashboard/src/presentation/widget/insight/configuration/DrillTargetType/DrillTargetTypeListItem.tsx (hunk 3 above)
Comment (phong-nguyen-duy) [reply libs/sdk-ui-dashboard/src/presentation/widget/insight/configuration/DrillTargetType/DrillTargetTypeListItem.tsx]: Done

File: libs/sdk-ui-dashboard/src/presentation/widget/insight/configuration/DrillTargets/AttributeHierarchyListItem.tsx (hunk 4 above)
Comment (phong-nguyen-duy) [reply libs/sdk-ui-dashboard/src/presentation/widget/insight/configuration/DrillTargets/AttributeHierarchyListItem.tsx]: Done

File: libs/sdk-ui-dashboard/src/presentation/widget/insight/configuration/DrillTargets/DrillTargetAttributeHierarchyItem.tsx (hunk 6 above)
Comment (phong-nguyen-duy) [reply libs/sdk-ui-dashboard/src/presentation/widget/insight/configuration/DrillTargets/DrillTargetAttributeHierarchyItem.tsx:71]: Hi @xMort, What do you think about this case? It isn't mentioned in mockup.

File: libs/sdk-ui-dashboard/src/presentation/widget/insight/configuration/DrillTargets/AttributeHierarchyListItem.tsx (hunk 7 above)
Comment (phong-nguyen-duy) [reply libs/sdk-ui-dashboard/src/presentation/widget/insight/configuration/DrillTargets/AttributeHierarchyListItem.tsx:65]: Nothing special, I just want to separate the responsibilities of these components.

File: libs/sdk-ui-dashboard/src/presentation/widget/insight/configuration/DrillTargets/DrillTargetAttributeHierarchyItem.tsx (hunk 6 above)
Comment (khanhhuynh-gooddata) [reply libs/sdk-ui-dashboard/src/presentation/widget/insight/configuration/DrillTargets/DrillTargetAttributeHierarchyItem.tsx:71]: BTW, if we display the empty message in this case, that means the dropdown will not render, and users cannot edit the attribute hierarchy.

cc @phong-nguyen-duy , @xMort

Comment (xMort) [reply libs/sdk-ui-dashboard/src/presentation/widget/insight/configuration/DrillTargets/DrillTargetAttributeHierarchyItem.tsx:71]: I think @khanhhuynh-gooddata is right.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/16869/ : SUCCESS in 3s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/18304/ : FAILURE in 3m 15s

File: libs/sdk-ui-dashboard/src/presentation/widget/insight/configuration/DrillTargets/DrillTargetAttributeHierarchyItem.tsx (hunk 6 above)
Comment (phong-nguyen-duy) [reply libs/sdk-ui-dashboard/src/presentation/widget/insight/configuration/DrillTargets/DrillTargetAttributeHierarchyItem.tsx]: Done, the dropdown will show all hierarchies which contains the origin attribute and the items are not in the blacklist will be disabled

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/16870/ : SUCCESS in 3s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/18305/ : SUCCESS in 12m 47s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/16871/ : SUCCESS in 3s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/18306/ : SUCCESS in 12m 47s

**Review by xMort — APPROVED** [review]

**Comment by yenkins** [comment]
### Sonar scan result


More detail, see in https://sonarqube-gate.intgdc.com/dashboard?id=gooddata-ui-sdk-gate-PR4339

To scan for vulnerabilities in dependencies and run unit tests (to get coverage report in sonar) please comment your PR with 'extended check sonar'.

**Comment by gdgate** [comment]
Build failed (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-sonarqube-scan-zuul-gate/3474/ : SUCCESS in 10m 00s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/18307/ : SUCCESS in 12m 55s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-cypress-bear-isolated-zuul-pipeline/1096/ : SUCCESS in 9m 48s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-merge-compliance-check/4639/ : SUCCESS in 9s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-fossa-scanning-zuul/1354/ : FAILURE in 1m 25s (non-voting)
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-cypress-tiger-isolated-zuul-pipeline/1036/ : UNSTABLE in 15m 01s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-compatibility-pipeline/533/ : SUCCESS in 5m 48s

**Comment by yenkins** [comment]
### Sonar scan result


More detail, see in https://sonarqube-gate.intgdc.com/dashboard?id=gooddata-ui-sdk-gate-PR4339

To scan for vulnerabilities in dependencies and run unit tests (to get coverage report in sonar) please comment your PR with 'extended check sonar'.

**Comment by gdgate** [comment]
Build failed (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-sonarqube-scan-zuul-gate/3475/ : SUCCESS in 12m 00s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/18308/ : SUCCESS in 12m 53s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-cypress-bear-isolated-zuul-pipeline/1097/ : UNSTABLE in 13m 04s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-merge-compliance-check/4640/ : SUCCESS in 12s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-fossa-scanning-zuul/1355/ : FAILURE in 1m 32s (non-voting)
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-cypress-tiger-isolated-zuul-pipeline/1037/ : SUCCESS in 14m 24s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-compatibility-pipeline/534/ : SUCCESS in 6m 51s

**Comment by yenkins** [comment]
### Sonar scan result


More detail, see in https://sonarqube-gate.intgdc.com/dashboard?id=gooddata-ui-sdk-gate-PR4339

To scan for vulnerabilities in dependencies and run unit tests (to get coverage report in sonar) please comment your PR with 'extended check sonar'.

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-sonarqube-scan-zuul-gate/3476/ : SUCCESS in 10m 27s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/18312/ : SUCCESS in 12m 48s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-cypress-bear-isolated-zuul-pipeline/1098/ : SUCCESS in 10m 25s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-merge-compliance-check/4641/ : SUCCESS in 10s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-fossa-scanning-zuul/1356/ : FAILURE in 1m 34s (non-voting)
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-cypress-tiger-isolated-zuul-pipeline/1039/ : SUCCESS in 15m 43s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-compatibility-pipeline/535/ : SUCCESS in 5m 56s

**Pull request merged**